File: 03/solution.ts
const input = await Bun.file(`${import.meta.dir}/input.txt`).text();
const lines = input.trim().split("\n");

// brute force
function part1() {
	let sum = 0;

	for (const line of lines) {
		const bank = line.split("").map((x) => Number.parseInt(x));
		let highestJoltage = 0;

		for (let a = 0; a < bank.length; a++) {
			for (let b = a + 1; b < bank.length; b++) {
				const j = Number.parseInt(`${bank[a]}${bank[b]}`);
				highestJoltage = Math.max(highestJoltage, j);
			}
		}

		sum += highestJoltage;
	}

	return sum;
}

// greedy pick largest digit while I have room
function part2() {
	let sum = 0;

	for (const line of lines) {
		const bank = line.split("").map((x) => Number.parseInt(x));
		let highestJoltage = "";
		let start = 0;

		while (highestJoltage.length < 12) {
			const remaining = 12 - highestJoltage.length;
			const end = bank.length - remaining;
			let highest = { j: 0, i: start };

			for (let i = start; i <= end; i++) {
				if (bank[i] > highest.j) {
					highest = { j: bank[i], i };
				}
			}

			highestJoltage += `${highest.j}`;
			start = highest.i + 1;
		}

		sum += Number.parseInt(highestJoltage);
	}

	return sum;
}

console.log("Part 1:", part1());
console.log("Part 2:", part2());

File: 03/solution.ml
open Base
open Stdio

let read_input () = In_channel.read_lines "03/input.txt"

(* brute force *)
let part1 lines =
  lines
  |> List.sum
       (module Int)
       ~f:(fun line ->
         let bank = String.to_list line |> List.map ~f:Char.get_digit_exn in
         let indexed = List.mapi bank ~f:(fun i d -> (i, d)) in
         List.cartesian_product indexed indexed
         |> List.filter_map ~f:(fun ((i, a), (j, b)) ->
             if i < j then Some ((a * 10) + b) else None)
         |> List.max_elt ~compare:Int.compare
         |> Option.value ~default:0)

(* greedy pick largest digit while I still have room *)
let part2 lines =
  lines
  |> List.sum
       (module Int)
       ~f:(fun line ->
         let bank = String.to_array line |> Array.map ~f:Char.get_digit_exn in
         let n = Array.length bank in
         let rec pick start needed acc =
           if needed = 0 then acc
           else
             let end_ = n - needed in
             let best_i, best_d =
               List.range start (end_ + 1)
               |> List.map ~f:(fun i -> (i, bank.(i)))
               |> List.max_elt ~compare:(fun (_, a) (_, b) -> Int.compare a b)
               |> Option.value_exn
             in
             pick (best_i + 1) (needed - 1) ((acc * 10) + best_d)
         in
         pick 0 12 0)

let () =
  let lines = read_input () in
  printf "Part 1: %d\n" (part1 lines);
  printf "Part 2: %d\n" (part2 lines)
